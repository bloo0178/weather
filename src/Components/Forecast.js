import React from 'react';
import axios from 'axios';
import moment from 'moment';


let APIkey = 'aadd2eb63b5dece08ab06ef70507cc13';

//Forecast example: http://api.openweathermap.org/data/2.5/forecast?zip=06831,us&APPID=aadd2eb63b5dece08ab06ef70507cc13&units=imperial&cnt=5

class Forecast extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            array: []
        }
    }

    componentDidMount() {
        axios.get(`http://api.openweathermap.org/data/2.5/forecast?` +
            `${this.props.location}&APPID=${APIkey}&units=${this.props.units}&cnt=50`)
            // ^^^ This will return a max of 40 objects with the free API. Cuts off the final data point on the last day.
            .then(res => {
                const data = (res.data.list);

                console.clear();
                console.log(data);
                //dateRange is used as the baseline to ensure displayed forecast doesn't exceed a 5-day window.
                var minDate = moment().format("YYYY-MM-DD");
                var maxDate = moment().add(5, 'days').format("YYYY-MM-DD");

                var newData = [];

                data.forEach((elem) => {
                    let elemDate = moment(elem.dt_txt, "YYYY-MM-DD HH:mm:ss").format("YYYY-MM-DD");
                    let elemTime = moment(elem.dt_txt, "YYYY-MM-DD HH:mm:ss").format("ha");
                    //Only show forecast for 5 days after today. This checks dataset to see if value is today or earlier.
                    if (moment(elemDate).isSameOrBefore(minDate)) return;
                    //This checks dataset to see if value is after 5 days from now.
                    if (moment(elemDate).isAfter(maxDate)) return; 
                    let temp = {
                        date: elemDate,
                        times: [{
                            time: elemTime,
                            temp: elem.main.temp,
                            description: elem.weather[0].main, 
                            icon: elem.weather[0].icon //not using currently
                        }]
                    }
                    if (!newData.some(x => x.date === elemDate)) {
                        newData.push(temp);
                    } else { 
                        for (let i = 0; i < newData.length; i++) { 
                            if (newData[i].date === elemDate) {
                                newData[i].times.push(temp.times[0]);
                                i+=newData.length;
                            }
                        }
                    }
                })
                this.setState({
                    array: newData
                })
            })
            .catch(err => {
                this.setState({ temp: "error" })
            })
    }

    render() {

        return this.state.array.map((day) => {
            return (
                <div className="Card-Wrapper">
                    <div>{moment(day.date, "YYYY-MM-DD").format("ddd MMM D")}</div>
                    {day.times.map((time) => {
                        return (
                            <div>{time.time + ": " +
                                time.temp + this.props.tempUnit +
                                " " + time.description}
                            </div>
                        )
                    })}
                </div>
            )
        })
    }
}


export default Forecast; 