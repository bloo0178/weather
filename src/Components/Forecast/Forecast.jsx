import React from "react";
import moment from "moment";
import getForecast from "../../api/forecastAPI";
import ForecastDayCard from "./ForecastDayCard/ForecastDayCard";
import styles from "./Forecast.module.css";
import SwipeableViews from "react-swipeable-views";

class Forecast extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      array: []
    };
  }

  componentDidUpdate = prevProps => {
    const { location, units } = this.props;
    if (location !== prevProps.location) {
      getForecast(location, units, newData => {
        this.setState({
          array: newData
        });
      });
    }
  };

  componentDidMount = () => {
    const { location, units } = this.props;
    getForecast(location, units, data => {
      console.log(data);
      this.setState({
        array: data
      });
    });
  };

  render() {
    return (
      <div className={styles.Container}>
        <SwipeableViews enableMouseEvents >
        {this.state.array.map(day => {
          return (
            <div className={styles.Cards} key={day.date}>
              <ForecastDayCard
                date={moment(day.date, "YYYY-MM-DD").format("ddd MMM D")}
                times={day.times}
                tempUnit={this.props.tempUnit}
              />
            </div>
          );
        })}
        </SwipeableViews>
      </div>
    );
  }
}

export default Forecast;
