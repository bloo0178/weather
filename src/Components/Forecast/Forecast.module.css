.Container {
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    box-sizing: border-box;
}

.Cards {
    display: flex; 
    flex-direction: column;
    justify-content: center;
    align-items: center;
}

.Test {
    height: 25rem;
    width: 25rem;
    background-color: orange;
}