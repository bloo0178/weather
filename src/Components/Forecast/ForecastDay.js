import React from 'react';
import { Collapse, Button, Table } from 'reactstrap';

class ForecastDay extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            collapse: false
        }
    }

    toggle = () => {
        this.setState({
            collapse: !this.state.collapse
        })
    }

    render() {
        return (
            <div className="Forecast-Wrapper">
                <span className="Forecast-Date">
                    {this.props.date}
                    <Button color="secondary" size="sm" onClick={this.toggle}>toggle</Button>
                </span>
                <Collapse isOpen={!this.state.collapse}>
                <Table striped size="sm" >
                    <tbody >
                    {this.props.times.map((time) => {
                        return (
                            <tr >
                                <td className="align-middle">{time.time}</td>
                                <td className="align-middle"><img src={time.icon} /></td>
                                <td className="align-middle">{time.temp + ' ' + this.props.tempUnit}</td>
                                <td className="align-middle">{time.description}</td>
                            </tr>
                        )
                    })}
                    </tbody>
                </ Table>
                </Collapse>
            </div>
        )
    }


}

export default ForecastDay;

