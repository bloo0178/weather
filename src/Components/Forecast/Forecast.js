import React from 'react';
import moment from 'moment';
import getForecast from './Forecast-API'
import ForecastDay from './ForecastDay';

class Forecast extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            array: [], 
        }
    }

    componentDidUpdate = (prevProps) => {
        if (this.props.location !== prevProps.location) {
            getForecast(this.props.location, this.props.units, (newData) => {
                this.setState({
                    array: newData
                })
            });
        }
    }

    componentDidMount = () => {
        getForecast(this.props.location, this.props.units, 
            (data) => {
                console.log(data);
                this.setState({
                    array: data
                })
            })
    }

    render() {
        
        return this.state.array.map((day) => {
            return (
                <ForecastDay 
                    date={moment(day.date, "YYYY-MM-DD").format("ddd MMM D")}
                    times={day.times}
                    tempUnit={this.props.tempUnit}
                />
            )
        })
    }
}


export default Forecast; 