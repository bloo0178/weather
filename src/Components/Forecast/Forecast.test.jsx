import React from "react";
import Forecast from "./Forecast";
import { shallow } from "enzyme";

let forecastData = [{ date: "2019-05-13", times: ["1", "2"] }];

describe("<Forecast />", () => {
	test("renders correctly", () => {
		let wrapper = shallow(
			<Forecast forecastData={forecastData} tempUnit="F" />
		);
		expect(wrapper).toMatchSnapshot();
	});
});
