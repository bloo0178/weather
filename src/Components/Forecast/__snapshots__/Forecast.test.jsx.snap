// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Forecast /> renders correctly 1`] = `
<div
  className="Container"
>
  <ReactSwipableView
    animateHeight={false}
    animateTransitions={true}
    axis="x"
    disableLazyLoading={false}
    disabled={false}
    enableMouseEvents={true}
    hysteresis={0.6}
    ignoreNativeScroll={false}
    index={0}
    resistance={false}
    springConfig={
      Object {
        "delay": "0s",
        "duration": "0.35s",
        "easeFunction": "cubic-bezier(0.15, 0.3, 0.25, 1)",
      }
    }
    threshold={5}
  >
    <div
      className="Cards"
      key="2019-05-13"
    >
      <ForecastDayCard
        date="Mon May 13"
        tempUnit="F"
        times={
          Array [
            "1",
            "2",
          ]
        }
      />
    </div>
  </ReactSwipableView>
</div>
`;
